lib/itertools/stream/stream.go

```diff
@@ -0,0 +1,459 @@
+// be optionally provided which will be run on close. If wrapping a
+// paginated API, consider using PageFunc instead.
+func Func[T any](fn func() (T, error)) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for {
+			item, err := fn()
+			if err != nil {
+				if !errors.Is(err, io.EOF) {
+					var zero T
+					yield(zero, trace.Wrap(err))
+				}
+				return
+			}
+
+			if !yield(item, nil) {
+				return
+			}
+		}
+	}
+}
+
+// Collect aggregates a stream into a slice. If an error is hit, the
+// items observed thus far are still returned, but they may not represent
+// the complete set.
+func Collect[T any](stream Stream[T]) ([]T, error) {
+	var c []T
+	for item, err := range stream {
+		if err != nil {
+			return c, err
+		}
+		c = append(c, item)
+	}
+	return c, nil
+}
+
+// CollectPages aggregates a paginated stream into a slice. If an error
+// is hit, the pages observed thus far are still returned, but they may not
+// represent the complete set.
+func CollectPages[T any](stream Stream[[]T]) ([]T, error) {
+	var c []T
+	for page, err := range stream {
+		if err != nil {
+			return c, err
+		}
+		c = append(c, page...)
+	}
+	return c, nil
+}
+
+// FilterMap maps a stream of type A into a stream of type B, filtering out
+// items when fn returns false.
+func FilterMap[A, B any](stream Stream[A], fn func(A) (B, bool)) Stream[B] {
+	return func(yield func(B, error) bool) {
+		for item, err := range stream {
+			if err != nil {
+				var zero B
+				yield(zero, trace.Wrap(err))
+				return
+			}
+
+			mapped, ok := fn(item)
+			if !ok {
+				continue
+			}
+
+			if !yield(mapped, nil) {
+				return
+			}
+		}
+	}
+}
+
+// MapWhile maps a stream of type A into a stream of type B, halting early
+// if fn returns false.
+func MapWhile[A, B any](stream Stream[A], fn func(A) (B, bool)) Stream[B] {
+	return func(yield func(B, error) bool) {
+		for item, err := range stream {
+			if err != nil {
+				var zero B
+				yield(zero, trace.Wrap(err))
+				return
+			}
+
+			mapped, ok := fn(item)
+			if !ok {
+				return
+			}
+
+			if !yield(mapped, nil) {
+				return
+			}
+		}
+	}
+}
+
+// Chain joins multiple streams in order, fully consuming one before moving to the next.
+func Chain[T any](streams ...Stream[T]) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for _, stream := range streams {
+			for item, err := range stream {
+				if err != nil {
+					var zero T
+					yield(zero, trace.Wrap(err))
+					return
+				}
+
+				if !yield(item, nil) {
+					return
+				}
+			}
+		}
+	}
+}
+
+// Chunks breaks a stream into chunks of a fixed size. The last chunk may be smaller
+// than the specified size. Zero/negative values of size result in an empty stream.
+func Chunks[T any](stream Stream[T], size int) Stream[[]T] {
+	if size < 1 {
+		return Empty[[]T]()
+	}
+	return func(yield func([]T, error) bool) {
+		var chunk []T
+		for item, err := range stream {
+			if err != nil {
+				yield(nil, trace.Wrap(err))
```

Comment:
Should we yield the items that are left in the last chunk before yielding the error?

Reply:
I don't like the idea of yielding a partial chunk in any context other than as the final chunk after a successfully completed stream.  I think it keeps the combinator easier to reason about when thats the only case where a partial chunk might be yielded, and I think short-circuiting on error is generally a desirable trait that we should stick to for this package.  If you don't want short-circuiting on error, I think using generic `iter.Seq2` combinators might be better.